ibm/resource_ibm_network_gateway.go

```diff
@@ -0,0 +1,857 @@
+			}
+
+			if bm.ActiveTransactionCount != nil && *bm.ActiveTransactionCount == 0 {
+				return bm, "idle", nil
+			}
+			return bm, "active", nil
+
+		},
+		Timeout:        24 * time.Hour,
+		Delay:          10 * time.Second,
+		MinTimeout:     1 * time.Minute,
+		NotFoundChecks: 24 * 60,
+	}
+
+	return stateConf.WaitForState()
+}
+
+// Network gateways or Bare metal creation does not return a  object with an Id.
+// Have to wait on provision date to become available on server that matches
+// hostname and domain.
+// http://sldn.softlayer.com/blog/bpotter/ordering-bare-metal-servers-using-softlayer-api
+func waitForNetworkGatewayProvision(d *datatypes.Hardware, meta interface{}) (interface{}, error) {
+	hostname := *d.Hostname
+	domain := *d.Domain
+	log.Printf("Waiting for Gateway (%s.%s) to be provisioned", hostname, domain)
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"retry", "pending"},
+		Target:  []string{"provisioned"},
+		Refresh: func() (interface{}, string, error) {
+			service := services.GetAccountService(meta.(ClientSession).SoftLayerSession())
+			bms, err := service.Filter(
+				filter.Build(
+					filter.Path("hardware.hostname").Eq(hostname),
+					filter.Path("hardware.domain").Eq(domain),
+				),
+			).Mask("id,provisionDate").GetHardware()
+			if err != nil {
+				return false, "retry", nil
+			}
+
+			if len(bms) == 0 || bms[0].ProvisionDate == nil {
+				return datatypes.Hardware{}, "pending", nil
+			} else {
+				return bms[0], "provisioned", nil
+			}
+		},
+		Timeout:        24 * time.Hour,
+		Delay:          10 * time.Second,
+		MinTimeout:     1 * time.Minute,
+		NotFoundChecks: 24 * 60,
+	}
+
+	return stateConf.WaitForState()
+}
+
+func expandVlans(configured []interface{}) []datatypes.Network_Gateway_Vlan {
+	vlans := make([]datatypes.Network_Gateway_Vlan, 0, len(configured))
+	for _, lRaw := range configured {
+		data := lRaw.(map[string]interface{})
+		p := &datatypes.Network_Gateway_Vlan{}
+		if v, ok := data["networkVlanID"]; ok && v.(int) != 0 {
+			p.NetworkVlanId = sl.Int(v.(int))
+		}
+		if v, ok := data["bypass"]; ok {
+			p.BypassFlag = sl.Bool(v.(bool))
+		}
+		if v, ok := data["networkGatewayID"]; ok && v.(int) != 0 {
+			p.NetworkGatewayId = sl.Int(v.(int))
+		}
+
+		vlans = append(vlans, *p)
+	}
+	return vlans
+}
+
+func resourceIBMNetworkGatewayVlanAssociate(d *schema.ResourceData, meta interface{}, vlanObject []datatypes.Network_Gateway_Vlan, id int) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	processingarray := vlanObject
+
+	for i := 0; i < len(processingarray); i++ {
+		bypass := *processingarray[i].BypassFlag
+		networkGatewayID := id
+		networkVlanID := *processingarray[i].NetworkVlanId
+		vlanObjectTemplate := datatypes.Network_Gateway_Vlan{
+			BypassFlag:       &bypass,
+			NetworkGatewayId: &networkGatewayID,
+			NetworkVlanId:    &networkVlanID,
+		}
+		_, err := services.GetNetworkGatewayVlanService(sess).CreateObject(&vlanObjectTemplate)
```

Comment:
I see there is way to pass multiple objects. Can we we that instead https://sldn.softlayer.com/reference/services/softlayer_network_gateway_vlan/createobjects?